### ISSUE TITLE
[HIPIFY] Issue running hipify-clang  - unit32_t

### ISSUE THREAD

**Issue by uahsaj** [ROCm-Developer-Tools/HIPIFY#92]
I am having an issue with running hipify-clang.  I have cloned the latest and built, downloaded and install CUDA 10.2, and downloaded LLVM+Clang 10.0.
First I get the warning :
warning: Unknown CUDA version 10.2. Assuming the latest supported version 10.1

I thought it was supposed to be compatible with 10.2?
Next when it tries to run and I am using the square.cu example, I get:

In file included from <built-in>:1:
In file included from /home/drew/hip/HIP/hipify-clang/dist/include/__clang_cuda_runtime_wrapper.h:48:
/usr/local/cuda-10.2/include/cuda.h:59:9: error: unknown type name 'uint32_t'
typedef uint32_t cuuint32_t;

Any ideas?

**Comment by incardon** [comment]
Same problem is there a work around in the meanwhile ?

**Comment by incardon** [comment]
Like it works with CUDA X.Y.X + LLVM X.Y.Z

**Comment by incardon** [comment]
... Well it look like for me the work around is to use a different installation folder for hipify different from the standard. 

It look like infact that hipify install in /usr/local/include headers that confuse clang++ at the level that after make install clang++ is not able to work anymore. Compiling even a simple program where i include just iostresm result in

In file included from main.cpp:1:
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/iostream:39:
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/ostream:38:
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/ios:40:
In file included from /usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/bits/char_traits.h:504:
/usr/lib/gcc/x86_64-linux-gnu/8/../../../../include/c++/8/cstdint:48:11: error: no member named 'int8_t' in the global namespace; did you mean 'wint_t'?
  using ::int8_t;
        ~~^
  

After creating a different install folder everything went back to normal

**Comment by incardon** [comment]
Nope ... actually does not solve the above. It only solve the broken clang++ ... so forget about my answer ... still no solution

**Comment by incardon** [comment]
Ok found a work around ... for me. Is on the same line of the previous discovery. Hipify add some headers that confuse clang the hipify-clang is used, the guilty headers are

HIPIFY_install/include/stdint.h
HIPIFY_install/include/cuda_wrappers/algorithm  
HIPIFY_install/include/cuda_wrappers/complex
HIPIFY_install/include/cuda_wrappers/new

renamed to file_broken

and than hipify-clang started to work on my machine

**Comment by emankov** [comment]
Clang system CUDA wrapping includes are implemented in a hacky way. Those clang header files are shipped with clang to compile CUDA by clang, and clang-based tools (like hipify-clang) as well. By installing hipify-clang those clang header files are being copied into `include` subfolder same level as hipify-clang executable. When running hipify-clang searches those header files there: 

`.../hipify-clang/dist/include/cuda_wrappers`
`.../hipify-clang/dist/include`

And clang searches in its own paths as well:

`.../lib/clang/10.0.0`
`.../lib/clang/10.0.0/include/cuda_wrappers`

Those paths are combined in every hipify-clang resulting command line.

Here is the root cause for conflict and errors reported due to hacky implementation of those header files. The header files are exactly the same, but their implementation leads to unobvious errors like reported. 

That is why the reported error happens only with installed hipify-clang when clang is also presented at its initial place.
 So currently, there are options for correctly running hipify-clang, being built from scratch:
1. Do not install hipify-clang - run it from its build folder - it should work with clang headers.
2. Rename source LLVM root folder - installed hipify-clang should work with its own clang headers (actually that was the aim for its installing - running hipify-clang on the systems without LLVM/clang).
3. Use hipify-clang from the package.

**The given explanation doesn't cancel the bug existence. I'll let you know when the fix becomes available.**

**Comment by emankov** [comment]
@uahsaj, @incardon, please check on your side, as it was fixed with #127.

**Comment by emankov** [comment]
Closing as fixed by #127

**Comment by misyntropy** [comment]
I am having the very same problem (seemingly) even though it should be fixed according to this issue 5 years later. I am providing `--cuda-path` but its refusing to see it. Please see my log and let me know if I have a different problem. In that case I will post a new issue. 

I am on Fedora 42, and have clang 17, 19 (libs only as a dependency to this package), and 20 installed for sure.

````
milind@number9-desktop-linux:~/rocmnotes$ hipify-clang softmax-gpucpu.cu --clang-resource-directory=/usr/lib/clang/19 --cuda-path=./cuda-12.5 -v
clang version 19.1.7 (Fedora 19.1.7-13.fc42)
Target: x86_64-redhat-linux-gnu
Thread model: posix
InstalledDir: /usr/bin
System configuration file directory: /etc/clang19/
Found candidate GCC installation: /usr/bin/../lib/gcc/x86_64-redhat-linux/15
Selected GCC installation: /usr/bin/../lib/gcc/x86_64-redhat-linux/15
Candidate multilib: .;@m64
Candidate multilib: 32;@m32
Selected multilib: .;@m64
Found HIP installation: /usr, version 6.3.42133
error: cannot find CUDA installation; provide its path via '--cuda-path', or pass '-nocudainc' to build without CUDA includes
clang Invocation:
 "/usr/bin/clang-tool" "-cc1" "-triple" "x86_64-redhat-linux-gnu" "-aux-triple" "nvptx64-nvidia-cuda" "-fsyntax-only" "-disable-free" "-clear-ast-before-backend" "-disable-llvm-verifier" "-discard-value-names" "-main-file-name" "softmax-gpucpu.cu-9254bb.hip" "-mrelocation-model" "static" "-mframe-pointer=all" "-fmath-errno" "-ffp-contract=on" "-fno-rounding-math" "-mconstructor-aliases" "-funwind-tables=2" "-target-cpu" "x86-64" "-tune-cpu" "generic" "-debugger-tuning=gdb" "-fdebug-compilation-dir=/home/milind/rocmnotes" "-v" "-fcoverage-compilation-dir=/home/milind/rocmnotes" "-resource-dir" "/usr/lib/clang/19" "-internal-isystem" "/usr/lib/clang/19/include/cuda_wrappers" "-I" "/home/milind/rocmnotes" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/x86_64-redhat-linux" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/backward" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/x86_64-redhat-linux" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/backward" "-internal-isystem" "/usr/lib/clang/19/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../x86_64-redhat-linux/include" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-internal-isystem" "/usr/lib/clang/19/include" "-internal-isystem" "/usr/local/include" "-internal-isystem" "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../x86_64-redhat-linux/include" "-internal-externc-isystem" "/include" "-internal-externc-isystem" "/usr/include" "-Wno-pragma-once-outside-header" "-std=c++14" "-fdeprecated-macro" "-ferror-limit" "19" "-fgnuc-version=4.2.1" "-fskip-odr-check-in-gmf" "-fcxx-exceptions" "-fexceptions" "-fcolor-diagnostics" "-faddrsig" "-D__GCC_HAVE_DWARF2_CFI_ASM=1" "-x" "cuda" "/tmp/softmax-gpucpu.cu-9254bb.hip"

clang -cc1 version 19.1.7 based upon LLVM 19.1.7 default target x86_64-redhat-linux-gnu
ignoring nonexistent directory "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../x86_64-redhat-linux/include"
ignoring nonexistent directory "/include"
ignoring nonexistent directory "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../x86_64-redhat-linux/include"
ignoring nonexistent directory "/include"
ignoring duplicate directory "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15"
ignoring duplicate directory "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/x86_64-redhat-linux"
ignoring duplicate directory "/usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/backward"
ignoring duplicate directory "/usr/lib/clang/19/include"
ignoring duplicate directory "/usr/local/include"
ignoring duplicate directory "/usr/include"
#include "..." search starts here:
#include <...> search starts here:
 /home/milind/rocmnotes
 /usr/lib/clang/19/include/cuda_wrappers
 /usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15
 /usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/x86_64-redhat-linux
 /usr/bin/../lib/gcc/x86_64-redhat-linux/15/../../../../include/c++/15/backward
 /usr/lib/clang/19/include
 /usr/local/include
 /usr/include
End of search list.
/tmp/softmax-gpucpu.cu-9254bb.hip:5:10: fatal error: 'cuda_runtime.h' file not found
    5 | #include <cuda_runtime.h>
      |          ^~~~~~~~~~~~~~~~
hipify-clang: /usr/include/clang/Basic/DirectoryEntry.h:136: RefTy &clang::FileMgr::MapEntryOptionalStorage<clang::FileEntryRef>::value() & [RefTy = clang::FileEntryRef]: Assertion `has_value()' failed.
 #0 0x00007f14f721407a llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) (/usr/lib64/llvm19/lib/libLLVM.so.19.1+0x21407a)
 #1 0x00007f14f7210f86 llvm::sys::RunSignalHandlers() (/usr/lib64/llvm19/lib/libLLVM.so.19.1+0x210f86)
 #2 0x00007f14f72147eb (/usr/lib64/llvm19/lib/libLLVM.so.19.1+0x2147eb)
 #3 0x00007f14f6a28bb0 __restore_rt (/lib64/libc.so.6+0x19bb0)
 #4 0x00007f14f6a8209c __pthread_kill_implementation (/lib64/libc.so.6+0x7309c)
 #5 0x00007f14f6a28a7e gsignal (/lib64/libc.so.6+0x19a7e)
 #6 0x00007f14f6a106d0 abort (/lib64/libc.so.6+0x16d0)
 #7 0x00007f14f6a10639 __assert_perror_fail (/lib64/libc.so.6+0x1639)
 #8 0x000055a175ae119b (/usr/bin/hipify-clang+0x9419b)
 #9 0x00007f14feda9f1a clang::Preprocessor::HandleHeaderIncludeOrImport(clang::SourceLocation, clang::Token&, clang::Token&, clang::SourceLocation, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1a9f1a)
#10 0x00007f14feda1c37 clang::Preprocessor::HandleIncludeDirective(clang::SourceLocation, clang::Token&, clang::detail::SearchDirIteratorImpl<true>, clang::FileEntry const*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1a1c37)
#11 0x00007f14feda27d7 clang::Preprocessor::HandleDirective(clang::Token&) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1a27d7)
#12 0x00007f14fed6ee41 clang::Lexer::LexTokenInternal(clang::Token&, bool) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x16ee41)
#13 0x00007f14fedea2ed clang::Preprocessor::Lex(clang::Token&) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1ea2ed)
#14 0x00007f14fee10ce3 (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x210ce3)
#15 0x00007f14fedf4be9 (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1f4be9)
#16 0x00007f14fee34a24 clang::Parser::ParseInnerNamespace(llvm::SmallVector<clang::Parser::InnerNamespaceInfo, 4u> const&, unsigned int, clang::SourceLocation&, clang::ParsedAttributes&, clang::BalancedDelimiterTracker&) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x234a24)
#17 0x00007f14fee33e23 clang::Parser::ParseNamespace(clang::DeclaratorContext, clang::SourceLocation&, clang::SourceLocation) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x233e23)
#18 0x00007f14fee0d897 clang::Parser::ParseDeclaration(clang::DeclaratorContext, clang::SourceLocation&, clang::ParsedAttributes&, clang::ParsedAttributes&, clang::SourceLocation*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x20d897)
#19 0x00007f14feefabc7 clang::Parser::ParseExternalDeclaration(clang::ParsedAttributes&, clang::ParsedAttributes&, clang::ParsingDeclSpec*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2fabc7)
#20 0x00007f14feef9126 clang::Parser::ParseTopLevelDecl(clang::OpaquePtr<clang::DeclGroupRef>&, clang::Sema::ModuleImportState&) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2f9126)
#21 0x00007f14fedf3ece clang::ParseAST(clang::Sema&, bool, bool) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x1f3ece)
#22 0x000055a175ada0f8 (/usr/bin/hipify-clang+0x8d0f8)
#23 0x00007f150161433b clang::FrontendAction::Execute() (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2a1433b)
#24 0x00007f1501576584 clang::CompilerInstance::ExecuteAction(clang::FrontendAction&) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2976584)
#25 0x00007f15018697f0 clang::tooling::FrontendActionFactory::runInvocation(std::shared_ptr<clang::CompilerInvocation>, clang::FileManager*, std::shared_ptr<clang::PCHContainerOperations>, clang::DiagnosticConsumer*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2c697f0)
#26 0x00007f150186946a clang::tooling::ToolInvocation::runInvocation(char const*, clang::driver::Compilation*, std::shared_ptr<clang::CompilerInvocation>, std::shared_ptr<clang::PCHContainerOperations>) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2c6946a)
#27 0x00007f1501867d79 clang::tooling::ToolInvocation::run() (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2c67d79)
#28 0x00007f150186b857 clang::tooling::ClangTool::run(clang::tooling::ToolAction*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2c6b857)
#29 0x00007f15018609de clang::tooling::RefactoringTool::runAndSave(clang::tooling::FrontendActionFactory*) (/usr/lib64/llvm19/lib/libclang-cpp.so.19.1+0x2c609de)
#30 0x000055a175af3988 (/usr/bin/hipify-clang+0xa6988)
#31 0x00007f14f6a12575 __libc_start_call_main (/lib64/libc.so.6+0x3575)
#32 0x00007f14f6a12628 __libc_start_main@GLIBC_2.2.5 (/lib64/libc.so.6+0x3628)
#33 0x000055a175a91bb5 (/usr/bin/hipify-clang+0x44bb5)
Aborted (core dumped)
````
